src/senaite/core/api/label.py

```diff
@@ -0,0 +1,277 @@
+    """
+    # Remove marker interface for AT and DX schema extension
+    noLongerProvides(obj, ICanHaveLabels)
+
+
+def set_obj_labels(obj, labels):
+    """Set the given labels to the object label storage
+    """
+    obj = api.get_object(obj)
+    # Always mark the object for schema extension
+    enable_label_schema(obj)
+    # always sort the labels before setting it to the storage
+    set_storage(obj, tuple(sorted(labels)))
+    # mark the object with the proper interface
+    if not labels:
+        noLongerProvides(obj, IHaveLabels)
+        uncatalog_object(obj)
+    else:
+        alsoProvides(obj, IHaveLabels)
+        catalog_object(obj)
+
+
+def add_obj_labels(obj, labels):
+    """Add one ore more labels to the object
+
+    :param obj: the object to label
+    :param labels: string or list of labels to add
+    :returns: The new labels
+    """
+    obj = api.get_object(obj)
+    # prepare the set of new labels
+    new_labels = set(get_obj_labels(obj))
+    for label in to_labels(labels):
+        new_labels.add(label)
+    # set the new labels
+    set_obj_labels(obj, new_labels)
+    return get_obj_labels(obj)
+
+
+def del_obj_labels(obj, labels):
+    """Remove labels from the object
+
+    :param obj: the object where the labels should be removed
+    :param labels: string or list of labels to remove
+    :returns: The new labels
+    """
+    obj = api.get_object(obj)
+    # prepare the set of new labels
+    new_labels = set(get_obj_labels(obj))
+    for label in to_labels(labels):
+        new_labels.discard(label)
+    # set the new labels
+    set_obj_labels(obj, new_labels)
+    return get_obj_labels(obj)
+
+
+def search_objects_by_label(label, **kw):
+    """Search for objects having one or more of the given labels
+
+    :param label: string or list of labels to search
+    :returns: Catalog results (brains)
+    """
+    labels = to_labels(label)
+    query = {
+        "labels": map(api.safe_unicode, labels),
+        "sort_on": "title",
+    }
+    # Allow to update the query with the keywords
+    query.update(kw)
+    return search(query, catalog=LABEL_CATALOG)
+
+
+def get_klass(portal_type):
+    """Returns the implementation class of the given portal type
+
+    :param portal_type: The portal_type to lookup the class for
+    :returns: Class object
+    """
+    portal_types = api.get_tool("portal_types")
+    fti = portal_types.getTypeInfo(portal_type)
+
+    if fti.product:
+        at_types = listTypes(fti.product)
+        for t in at_types:
+            if not t.get("portal_type") == portal_type:
+                continue
+            return t.get("klass")
+    else:
+        return resolveDottedName(fti.klass)
+
+
+def enable_labels_for_type(portal_type):
+    """Enable labels for all objects of the given type
+
+    :param portal_type: The portal_type to enable labeling
+    """
+    klass = get_klass(portal_type)
+    classImplements(klass, ICanHaveLabels)
```

Comment:
For AT types, this works properly, but for DX types you need to programmatically add the behavior. Something like follows:

```python
behavior_id = 'senaite.core.behavior.labels'
pt = api.get_tool("portal_types")
fti = pt.get(portal_type)
if behavior_id not in fti.behaviors:
    fti_behaviors = list(fti_behaviors)
    fti_behaviors.append(behavior_id)
    fti.behaviors = tuple(fti_behaviors)
```

Reply:
You might be right and I thought too complicated.
I actually implemented this adapter to hook in the extended schema if the object implements `ICanHaveLabels`:
https://github.com/senaite/senaite.core/blob/8259ed672392308b911a4008a2c1cb06b741e34b/src/senaite/core/behaviors/label.py#L25

My initial thought was, that only by marking the class I can apply he behavior to both existing and new objects accordingly.
But the class interface is not looked up when a new DX type is added over the UI and seems to inherit then the interface in the object, which seems to make it not possible to disable it again w/o touching every object.
But adding/removing the behavior to the FTI programmatically might just solve the issue. Let me check this...